pkg/http/server.go

```diff
@@ -0,0 +1,195 @@
+	DefaultReadTimeout = 30 * time.Second
+
+	// DefaultWriteTimeout - default timout to write data to accepted connection.
+	DefaultWriteTimeout = 30 * time.Second
+
+	// DefaultMaxHeaderBytes - default maximum HTTP header size in bytes.
+	DefaultMaxHeaderBytes = 1 * humanize.MiByte
+)
+
+var sslRequiredErrMsg = []byte("SSL required")
+
+// Server - extended http.Server supports multiple addresses to serve and enhanced connection handling.
+type Server struct {
+	http.Server
+	Addrs                  []string                            // addresses on which the server listen for new connection.
+	ShutdownTimeout        time.Duration                       // timeout used for graceful server shutdown.
+	TCPKeepAliveTimeout    time.Duration                       // timeout used for underneath TCP connection.
+	UpdateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	UpdateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	ErrorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+	listenerMutex          *sync.Mutex                         // to guard 'listener' field.
+	listener               *httpListener                       // HTTP listener for all 'Addrs' field.
+	inShutdownMutex        *sync.Mutex                         // to guard 'inShutdown' field
+	inShutdown             bool                                // indicates whether the server is in shutdown or not
+	requestCount           int32                               // counter holds no. of request in process.
+}
+
+// Start - start HTTP server
+func (srv *Server) Start() (err error) {
+	// Take a copy of server fields.
+	tlsConfig := srv.TLSConfig
+	readTimeout := srv.ReadTimeout
+	writeTimeout := srv.WriteTimeout
+	handler := srv.Handler
+
+	addrs := srv.Addrs
+	tcpKeepAliveTimeout := srv.TCPKeepAliveTimeout
+	updateBytesReadFunc := srv.UpdateBytesReadFunc
+	updateBytesWrittenFunc := srv.UpdateBytesWrittenFunc
+	errorLogFunc := srv.ErrorLogFunc
+
+	// Create new HTTP listener.
+	var listener *httpListener
+	listener, err = newHTTPListener(
+		addrs,
+		tlsConfig,
+		tcpKeepAliveTimeout,
+		readTimeout,
+		writeTimeout,
+		updateBytesReadFunc,
+		updateBytesWrittenFunc,
+		errorLogFunc,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Wrap given handler to do additional
+	// * return 403 (forbidden) with 'SSL required' error message if non-encrypted
+	//   request received for secured server.
+	// * return 503 (service unavailable) if the server in shutdown.
+	wrappedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&srv.requestCount, 1)
+		defer atomic.AddInt32(&srv.requestCount, -1)
+
+		// If TLS is configured and the request is sent as non-TLS, return 403 (forbidden).
+		if tlsConfig != nil && r.TLS == nil {
```

Comment:
Since TLS handshake will fail with a non-TLS client, will this block ever be executed? Or is not the case that the handshake always fails if server is TLS, and client is not?